### PR TITLE
OSU mvapich2 GDR 2.3.5 package.py update

### PR DESCRIPTION
**Pull request by apreifsteck**
My name is Austin Reifsteck. I am working to support the MVAPICH project at OSU. I have become one of the maintainers of our Spack deployments. I have created this PR to update the package.py file to support the MVAPICH2 GDR 2.3.5 release. I am new to Spack, so please let me know if you see anything amiss! Thank you.

### CODE REVIEW

File: var/spack/repos/builtin/packages/mvapich2-gdr/package.py [hunk 1]
```diff
@@ -17,10 +16,11 @@ class Mvapich2Gdr(AutotoolsPackage):
     """
 
     homepage = 'http://mvapich.cse.ohio-state.edu'
-    url      = 'http://mvapich.cse.ohio-state.edu/download/mvapich/spack-mirror/mvapich2-gdr/mvapich2-gdr-2.3.4.tar.gz'
+    url      = 'file:///home/packager/spack-related/tar-bundles/mvapich2-gdr/mvapich2-gdr-2.3.5.tar.gz'
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/mvapich2-gdr/package.py]: Can you revert this change? Using `file` means that people have to resort to manual downloading the tarball. Is that what you meant?

File: var/spack/repos/builtin/packages/mvapich2-gdr/package.py [hunk 2]
```diff
@@ -71,12 +71,33 @@ class Mvapich2Gdr(AutotoolsPackage):
         default=False
     )
 
+    variant(
+        'cuda',
+        description='Enable/Disable support for cuda',
+        default=True
+    )
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/mvapich2-gdr/package.py:79]: Does it make sense to depend on `CudaPackage` to enable cuda support? Same question for Rocm.

File: var/spack/repos/builtin/packages/mvapich2-gdr/package.py [hunk 3]
```diff
@@ -71,12 +71,33 @@ class Mvapich2Gdr(AutotoolsPackage):
...
     depends_on('libpciaccess@0.13.5', when=(sys.platform != 'darwin'))
     depends_on('libxml2@2.9.10')
-    depends_on('cuda@9.2.88:10.2.89')
+    depends_on('cuda@9.2.88:11.1.1', when='+cuda')
     depends_on('pmix@3.1.3', when='pmi_version=pmix')
 
+    #ROCm dependencies
+    rocm_variant = '+rocm'
+    rocm_ver = '3.9.0'
+    depends_on('hip@' + rocm_ver, when=rocm_variant)
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/mvapich2-gdr/package.py]:
```suggestion
    depends_on('hip@3.9.0', when='+rocm')
```

File: var/spack/repos/builtin/packages/mvapich2-gdr/package.py [hunk 4]
```diff
@@ -124,12 +153,7 @@ def process_manager_options(self):
                 '--with-pm=hydra',
                 '--with-pbs=/opt/pbs'
             ])
-            if '~mcast' in spec:
-                opts.append('--disable-mcast')
-            if '+core_direct' in spec:
-                opts.append('--with-core-direct')
-            if '+openacc' in spec:
-                opts.append('--enable-openacc')
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/mvapich2-gdr/package.py:132]: Can you comment why these conditional branches have been removed? Are the variants involved used anywhere else or should they be removed too?

**Review by alalazo — CHANGES_REQUESTED** [review]
I left a few comments. Let me know what you think.

File: var/spack/repos/builtin/packages/mvapich2-gdr/package.py (hunk 4 above)
Comment (apreifsteck) [reply var/spack/repos/builtin/packages/mvapich2-gdr/package.py:132]: Of course. If you look at line 123 in the update file, those same conditional branches were already present. I removed these because they were redundant.

File: var/spack/repos/builtin/packages/mvapich2-gdr/package.py (hunk 2 above)
Comment (apreifsteck) [reply var/spack/repos/builtin/packages/mvapich2-gdr/package.py:79]: I've been told that since gdr is only available through selectively available binaries, when a user does `spack buildcache list` they won't be able to see which hash is for a cuda build and which is for rocm. 
 @nithintsk has made a feature request to address this, it seems https://github.com/spack/spack/issues/17998

**Review by apreifsteck — COMMENTED** [review]
I think I've addressed everything you mentioned. Thank you for the careful review!

**Review by tldahlgren — CHANGES_REQUESTED** [review]
Please resolve the flake8 errors:

- var/spack/repos/builtin/packages/mvapich2-gdr/package.py:9: [E302] expected 2 blank lines, found 1
15
- var/spack/repos/builtin/packages/mvapich2-gdr/package.py:88: [W293] blank line contains whitespace
16
- var/spack/repos/builtin/packages/mvapich2-gdr/package.py:90: [E303] too many blank lines (2)
17
- var/spack/repos/builtin/packages/mvapich2-gdr/package.py:153: [W293] blank line contains whitespace
18
- var/spack/repos/builtin/packages/mvapich2-gdr/package.py:155: [E303] too many blank lines (2)

File: var/spack/repos/builtin/packages/mvapich2-gdr/package.py (hunk 2 above)
Comment (alalazo) [reply var/spack/repos/builtin/packages/mvapich2-gdr/package.py:79]: Ok, thanks for the pointer to #17998 which fell definitely off my radar! What I am asking though is not to remove or change the variant, but if it makes sense to inherit it via the `CudaPackage` base class (which does also other things that may be wanted / not wanted depending on the package).

Comment (apreifsteck) [reply var/spack/repos/builtin/packages/mvapich2-gdr/package.py:79]: Hmm I hadn't considered that. I am pretty much brand new to the Spack ecosystem and don't quite know the lay of the land. I'll take a look at the docs, but my naive question is that if that would be a problem, since there is also a rocm variant now? And would that cause problems, since (to my knowledge), they are mutually exclusive?

**Review by apreifsteck — COMMENTED** [review]
Flake8 issues fixed.

File: var/spack/repos/builtin/packages/mvapich2-gdr/package.py (hunk 2 above)
Comment (apreifsteck) [reply var/spack/repos/builtin/packages/mvapich2-gdr/package.py:79]: Upon looking at the docs, I think it is a good suggestion, but it does not seem appropriate right now for a couple of reasons. The first is that for GDR I think we want building with cuda to be enabled by default (although I guess I could dig into it and find a way to override `CudaPackage`'s default behaivor). I also do not think we do anything with CUDA architectures at  this point, but perhaps @nithintsk can comment.

**Review by alalazo — APPROVED** [review]

**Comment by nithintsk** [comment]
@tldahlgren   Request you review and approval to merge these changes.

**Review by tldahlgren — APPROVED** [review]
Flake8 errors fixed and already approved by alalazo.

**Comment by tldahlgren** [comment]
Thank you for your contribution!
